protocol/communities/community_admin_event.go

```diff
@@ -0,0 +1,346 @@
+		CommunityId:    o.ID(),
+		Type:           protobuf.CommunityAdminEvent_COMMUNITY_MEMBER_KICK,
+		MemberToAction: pubkey,
+	}
+}
+
+func (o *Community) ToCommunityEditAdminEvent() *protobuf.CommunityAdminEvent {
+	return &protobuf.CommunityAdminEvent{
+		Clock:       o.Clock(),
+		CommunityId: o.ID(),
+		Type:        protobuf.CommunityAdminEvent_COMMUNITY_EDIT,
+		CommunityConfig: &protobuf.CommunityConfig{
+			Identity:      o.Identity(),
+			Permissions:   o.Permissions(),
+			AdminSettings: o.AdminSettings(),
+			IntroMessage:  o.IntroMessage(),
+			OutroMessage:  o.OutroMessage(),
+			Tags:          o.TagsRaw(),
+		},
+	}
+}
+
+func (o *Community) ToCommunityBecomeMemberTokenPermissionChangeAdminEvent() *protobuf.CommunityAdminEvent {
+	tokenPermissions := make(map[string]*protobuf.CommunityTokenPermission)
+	for _, tp := range o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER) {
+		tokenPermissions[tp.Id] = tp
+	}
+	return &protobuf.CommunityAdminEvent{
+		Clock:            o.Clock(),
+		CommunityId:      o.ID(),
+		Type:             protobuf.CommunityAdminEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_CHANGE,
+		TokenPermissions: tokenPermissions,
+	}
+}
+
+func (o *Community) ToCommunityBecomeMemberTokenPermissionDeleteAdminEvent() *protobuf.CommunityAdminEvent {
+	tokenPermissions := make(map[string]*protobuf.CommunityTokenPermission)
+	for _, tp := range o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER) {
+		tokenPermissions[tp.Id] = tp
+	}
+	return &protobuf.CommunityAdminEvent{
+		Clock:            o.Clock(),
+		CommunityId:      o.ID(),
+		Type:             protobuf.CommunityAdminEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_DELETE,
+		TokenPermissions: tokenPermissions,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinAcceptAdminEvent(changes *CommunityAdminEventChanges) *protobuf.CommunityAdminEvent {
+	return &protobuf.CommunityAdminEvent{
+		Clock:                  o.Clock(),
+		CommunityId:            o.ID(),
+		Type:                   protobuf.CommunityAdminEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT,
+		MembersAdded:           changes.MembersAdded,
+		AcceptedRequestsToJoin: changes.AcceptedRequestsToJoin,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinRejectAdminEvent(changes *CommunityAdminEventChanges) *protobuf.CommunityAdminEvent {
+	return &protobuf.CommunityAdminEvent{
+		Clock:                  o.Clock(),
+		CommunityId:            o.ID(),
+		Type:                   protobuf.CommunityAdminEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT,
+		RejectedRequestsToJoin: changes.RejectedRequestsToJoin,
+	}
+}
+
+func (o *Community) PatchCommunityDescriptionByAdminEvent(adminEvent *protobuf.CommunityAdminEvent) (*protobuf.CommunityDescription, error) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if adminEvent.Clock <= o.config.CommunityDescription.Clock {
+		return nil, errors.New("clock for admin event is outdated")
+	}
+
+	if adminEvent.Type == protobuf.CommunityAdminEvent_UNKNOWN {
```

Comment:
This is tricky, it makes the protocol non-upgradable, something worth thinking about (have to better look at the code to judge), but say we add a new event, clients will refuse to process it and return an error (which might be fine, as long as that doesn't prevent us to process further events).

Reply:
1 admin event == 1 msg, from my understanding, if there will be an error, we will proceed next msg

Reply:
cool, that's good